airflow/providers/common/sql/hooks/sql.py

```diff
@@ -30,13 +30,10 @@
 from airflow.utils.module_loading import import_string
 from airflow.version import version
 
-if TYPE_CHECKING:
-    from sqlalchemy.engine import CursorResult
 
-
-def fetch_all_handler(cursor: 'CursorResult') -> Optional[List[Tuple]]:
+def fetch_all_handler(cursor) -> Optional[List[Tuple]]:
     """Handler for DbApiHook.run() to return results"""
-    if cursor.returns_rows:
+    if cursor.description is not None:
         return cursor.fetchall()
```

Comment:
This doesn’t make sense. `description` only returns some information of the cursor and has nothing to do to whether the cursor returns data or not.

According to PEP 249, whether a cursor returns information can be checked by

```python
if cursor.rowcount is not None and cursor.rowcount >= 0
```

Reply:
This doesn't look true to me, I am using the following as reference:

- https://peps.python.org/pep-0249/#description
- https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=returns_#sqlalchemy.engine.CursorResult.returns_rows

Also:

```python
>>> import pymssql
>>> c = pymssql.connect(host, login, password)
>>> cur = c.cursor()
>>> cur.execute("SELECT SUSER_SNAME();")
>>> cur.rowcount
-1
>>> cur.description
(('', 1, None, None, None, None, None),)
>>> cur.execute("PRINT('1');")
>>> cur.rowcount
-1
>>> cur.description
```

Reply:
I am not sure how pymssql does things, but according to PEP 249, `description` does not offer the same functionality as SQLAlchemy’s `return_rows`. If `rowcount` does not either, you need to find another way that actually has a backing standard. Since DbApiHook should work for _all_ standard-compliant databases, we can’t rely on individual database behaviours, but must refer to the standard.

Reply:
@uranusjr Totally agree on the all standard-compliant part. However then the sqlalchemy's documentation is wrong then since for returns_rows it only mentions description. Do you have an example where description is not None and no rows where returned ?

Reply:
@uranusjr I find sqlalchemy notes on rowcount: https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=returns_#sqlalchemy.engine.CursorResult.rowcount very interesting

Reply:
Quoting sqlalchemy (aforementioned link):

- "Statements that use RETURNING may not return a correct rowcount."
- "Contrary to what the Python DBAPI says, it does not return the number of rows available from the results of a SELECT statement as DBAPIs cannot support this functionality when rows are unbuffered."

Reply:
```
about cursor.description https://peps.python.org/pep-0249/#description "This attribute will be None for operations that do not return rows or if the cursor has not had an operation invoked via the [.execute*()](https://peps.python.org/pep-0249/#id14) method yet."
```

This is indeed part of the standardm, so I do no see why we should not base the decision on that @uranusjr ? It's quite explicitly stated in the PEP that description is only present when there are some rows potentially to be returend.

Reply:
What I 'particularly` do not like about rowcount is this not about `-1`:


```
The attribute is -1 in case no [.execute*()](https://peps.python.org/pep-0249/#id14) 
has been performed on the cursor or the rowcount of the last operation 
is cannot be determined by the interface. [[7]](https://peps.python.org/pep-0249/#id46)

Note
Future versions of the DB API specification could redefine the latter case to have the object return None instead of -1.
```

I think just having the note indicate that we should avoid it, and there is absolutely no more guarantees rowcount gives us than description:

``
This attribute will be None for operations that do not return rows
or if the cursor has not had an operation
invoked via the [.execute*()](https://peps.python.org/pep-0249/#id14) method yet."
```

This is the same, only less ambiguous IMHO.